File: neuralnet.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "layers.c"


struct nn {
    struct layer *layers;
    int len;
}; // array of layers

double rand_double(double min, double max)  {
    double ans = min + (rand() / (RAND_MAX / (max - min)));
    return ans;
}

/*
use network_init(..) in create_nn(..)
*/

void network_init(struct nn *network, int randomized) {
    struct nn *temp_net;
    temp_net = (struct nn *) network;

    if (randomized == 1) {
        // generate network with weights and biases
        // set to random numbers between -1 and 1
        for (int i = 0; i < temp_net->len; i++) {
            init_weights(&temp_net->layers[i]);
            init_biases(&temp_net->layers[i]);

            for (int j = 0; j < temp_net->layers[i].layer_len; j++) {
                temp_net->layers[i].biases[j] = rand_double(-1, 1);
                for (int k = 0; k < temp_net->layers[i].input_len; k++) {
                    temp_net->layers[i].weights[j][k] = rand_double(-1, 1);
                }
            }
        }
    } else if (randomized == 0) {
        // generate network with weights and biases of 0.0
        for (int i = 0; i < temp_net->len; i++) {
            init_weights(&temp_net->layers[i]);
            init_biases(&temp_net->layers[i]);
        }
    } else {
        printf("Invalid randomized argument of '%i\n'", randomized);
    }
}

void layer_init(struct nn *network, int layer_num, int randomized) {
    struct nn *temp_net;
    temp_net = (struct nn *) network;

    if (layer_num >= temp_net->len || layer_num < 0) {
        printf("Invalid layer number\n");
        return;
    }

    if (randomized == 1) {
        // generate layer with weights and biases
        // set to random numbers between -1 and 1
        init_weights(&temp_net->layers[layer_num]);
        init_biases(&temp_net->layers[layer_num]);

        for (int i = 0; i < temp_net->layers[layer_num].layer_len; i++) {
            temp_net->layers[layer_num].biases[i] = rand_double(-1, 1);
            for (int j = 0; j < temp_net->layers[layer_num].input_len; j++) {
                temp_net->layers[layer_num].weights[i][j] = rand_double(-1, 1);
            }
        }
    } else if (randomized == 0) {
        // generate network with weights and biases of 0.0
        init_weights(&temp_net->layers[layer_num]);
        init_biases(&temp_net->layers[layer_num]);
    } else {
        printf("Invalid randomized argument of '%i\n'", randomized);
    }
}

struct nn create_nn(char *arc) {
    // create nn based on string
    // structure:
    // (int x)(char activation)|repeat
    // example
    // 3i|4r|6r|6r|2s|
    // nn with 4 layers, 3 input size, 4 relu neurons, 
    // 6 relu neurons, 6 relu neurons, 2 sigmoid neurons

    int arc_size = (int) strlen(arc);

    struct nn new_nn;
    new_nn.len = -1; // ignore the first and last as they dont indicate layers
    for (int i = 0; i < arc_size; i++) {
        if (arc[i] == '|') {
            new_nn.len++;
        }
    }
    new_nn.layers = malloc((new_nn.len - 1) * sizeof(struct layer));

    char *temp_string;
    temp_string = malloc(1 * sizeof(char));
    int last_len;
    int layer_count;
    for (int i = 0; i < arc_size; i++) {
        if (arc[i] >= '0' && arc[i] <= '9') {
            temp_string = (char *) realloc(temp_string, ((int) strlen(temp_string) + 1) * sizeof(char));
            strncat(temp_string, &arc[i], 1);
        } else if (arc[i] == '|') {
            if (arc[i-1] != 'i') {
                struct layer sheet;
                sheet.activation = arc[i-1];
                sheet.input_len = last_len;

                int layer_len = atoi(temp_string);
                sheet.layer_len = layer_len;

                last_len = layer_len;

                layer_count++;
                
                new_nn.layers[layer_count-1] = sheet;
            } else {
                last_len = atoi(temp_string);
            }
            memset(temp_string, 0, strlen(temp_string));
        }
    }

    return new_nn;
} 

void add_layer(struct nn *network, char *arc_part) {
    // test this and previous void funcs that edit structs
    struct nn *temp_net;
    temp_net = (struct nn *) network;
    
    // using realloc runs an error dont try that 
    struct layer *layers;
    layers = malloc(temp_net->len * sizeof(struct layer));
    for (int i = 0; i < temp_net->len; i++) {
        layers[i] = temp_net->layers[i];
    }
    temp_net->layers = layers;
    
    int arc_p_len = (int) strlen(arc_part);
    char *temp_string;
    temp_string = malloc(1 * sizeof(char));
    for (int i = 0; i < arc_p_len; i++) {
         if (arc_part[i] >= '0' && arc_part[i] <= '9') {
            temp_string = (char *) realloc(temp_string, ((int) strlen(temp_string) + 1) * sizeof(char));
            strncat(temp_string, &arc_part[i], 1);
        } else if (arc_part[i] == '|') {
            struct layer sheet;
            sheet.activation = arc_part[i-1];
            sheet.input_len = temp_net->layers[temp_net->len-1].layer_len;

            int layer_len = atoi(temp_string);
            sheet.layer_len = layer_len;
                
            temp_net->layers[temp_net->len] = sheet;
            
            memset(temp_string, 0, strlen(temp_string));
        }
    }

    temp_net->len++;
    // make sure this all writes to correct struct
}

double *feedforward(struct nn *network, double *inputs) {
    struct nn *temp_net;
    temp_net = (struct nn *) network;

    double *ans = inputs;
    // get prediction based on inputs
    for (int i = 0; i < temp_net->len; i++) {
        ans = output(&temp_net->layers[i], ans);
    }

    return ans;
}

File: neuralnet.py
import re
import wrapper


def checkValidArc(string):
    valid_arc = r'^(\d+i\|)(\d+[a-hj-zA-Z]\|)+$'
    if re.search(valid_arc, string):
        return
    else:
        raise ValueError(f'{string} is an invalid architecture')

def checkValidArcP(string):
    valid_arc_p = r'^(\d+[a-hj-zA-Z]\|)$'
    if re.search(valid_arc_p, string):
        return
    else:
        raise ValueError(f'{string} is an invalid architecture')

class Layer:
    def __init__(self, arcp=None, randomized=False):
        if arcp is not None:
            checkValidArcP(arcp)
            # use architecture
            if randomized:
                # random init
                pass
            else:
                pass
        else:
            # init empty layer
            pass
        raise NotImplementedError()
    
    def getLayerStructFromArc(self, arcp):
        raise NotImplementedError()
    
    def getLayerStructFromSpecs(self, input_len, layer_len, activation):
        raise NotImplementedError()

    def __repr__(self):
        # +--------------+
        # | activation:  |
        # | layer input: |
        # | layer len:   |
        # +--------------+
        rep = ''
        keys = {'activation: ': self.layer.activation,
                'layer input: ': self.layer.layer_input,
                'layer len: ': self.layer.layer_len}

        max_len = max([i + str(j) for i, j in keys.items()], key=len)
        bounds = '+' + ((max_len + 2) * '-') + '+'
        rep += bounds + '\n'
        for i in keys.items():
            rep += f'| {i[0]}{i[1]} |\n'
        rep += bounds + '\n'

        return rep

        # needs testing

class NeuralNetwork:
    def __init__(self, arc=None, randomized=False):
        if arc is not None:
            checkValidArc(arc)
            # use architecture
            if randomized:
                # random init
                pass
            else:
                pass
        else:
            # init empty network
            pass
        raise NotImplementedError()

    def getNetworkStructFromArc(self, arc):
        raise NotImplementedError 

    def edit_activation(self, layer_number, layer_activation):
        # edit layer activation
        raise NotImplementedError()

    def add_layer(self, layer_activation=None, layer_length=None, layer_input_length=None, restr=None):
        if restr and all(i is None for i in [layer_activation, length, layer_input_length]):
            checkValidArcP(restr)
            # use restr directly
            pass
        elif restr is None and not all(i is None for i in [layer_activation, length, layer_input_length]):
            # use args
            pass
        else:
            raise ValueError()
        # check if layer is first
        # edit layer if it is first
        raise NotImplementedError()

    def feedforward(self, input, timed=False):
        # use feedforward from c wrapper
        raise NotImplementedError()

    def __repr__(self):
        # format structure:
        # +--------------+
        # | nn length:   |
        # +--------------+
        # | activation:  |
        # | layer input: |
        # | layer len:   |
        # +--------------+
        # repeat until gone through all layers
        rep = '+' + ('-' * (len('nn length: ') + len(str(self.len)) + 2)) + '+\n'
        rep += f'| nn length: {self.len} |\n'
        for n in range(self.len):
            keys = {'activation: ': self.layer.activation,
                    'layer input: ': self.layer.layer_input,
                    'layer len: ': self.layer.layer_len}

            max_len = max([i + str(j) for i, j in keys.items()], key=len)
            bounds = '+' + ((max_len + 2) * '-') + '+'
            rep += bounds + '\n'
            for i in keys.items():
                rep += f'| {i[0]}{i[1]} |\n'
            rep += bounds + '\n'

        return rep

    def load(self, filename):
        raise NotImplementedError

    def save(self, filename):
        raise NotImplementedError